versioned_hdf5/subchunk_map.py

```diff
@@ -0,0 +1,308 @@
+import itertools
+from typing import Iterable, Iterator
+
+import cython
+import numpy as np
+from cython import Py_ssize_t
+from ndindex import (
+    BooleanArray,
+    ChunkSize,
+    Integer,
+    IntegerArray,
+    Slice,
+    Tuple,
+    ndindex,
+)
+
+
+@cython.cfunc
+def _ceiling(a: Py_ssize_t, b: Py_ssize_t) -> Py_ssize_t:
+    """Returns ceil(a/b)
+    """
+    return -(-a // b)
+
+
+@cython.cfunc
+def _smallest(x: Py_ssize_t, a: Py_ssize_t, m: Py_ssize_t) -> Py_ssize_t:
+    """Find the smallest integer y >= x where y = a + k*m for whole k's
+    Assumes 0 <= a <= x and m >= 1.
+
+    a                  x    y
+    | <-- m --> | <-- m --> |
+    """
+    n: Py_ssize_t = _ceiling(x - a, m)
+    return a + n * m
+
+
+@cython.cfunc
+def _subindex_chunk_slice(
+    c_start: Py_ssize_t,
+    c_stop: Py_ssize_t,
+    i_start: Py_ssize_t,
+    i_stop: Py_ssize_t,
+    i_step: Py_ssize_t,
+) -> slice:
+    common: Py_ssize_t = i_start % i_step
+
+    start: Py_ssize_t = max(c_start, i_start)
+    # Get the smallest lcm multiple of common that is >= start
+    start = _smallest(start, common, i_step)
+    # Finally, we need to shift start so that it is relative to index
+    start = (start - i_start) // i_step
+
+    stop: Py_ssize_t = min(c_stop, i_stop)
+    stop = _ceiling(stop - i_start, i_step) if stop > i_start else 0
+
+    return slice(start, stop, 1)
+
+
+@cython.cfunc
+def _subindex_slice_chunk(
+    s_start: Py_ssize_t,
+    s_stop: Py_ssize_t,
+    s_step: Py_ssize_t,
+    c_start: Py_ssize_t,
+    c_stop: Py_ssize_t,
+) -> slice:
+    common: Py_ssize_t = s_start % s_step
+
+    start: Py_ssize_t = max(s_start, c_start)
+    # Get the smallest step multiple of common that is >= start
+    start = _smallest(start, common, s_step)
+    # Finally, we need to shift start so that it is relative to index
+    start -= c_start
+
+    stop: Py_ssize_t = max(0, min(s_stop, c_stop) - c_start)
+
+    # This is the same, in the special case we're in, to
+    #     return Slice(start, stop, s_step).reduce(d).raw
+    # It's reimplemented here for speed.
+    # assert 0 <= start < stop <= d
+    step: Py_ssize_t = s_step
+    if start + step >= stop:
+        stop, step = start + 1, 1  # Indexes 1 element
+    else:
+        stop -= (stop - start - 1) % step
+        if stop - start == 1:
+            step = 1  # Indexes 1 element
+    return slice(start, stop, step)
```

Comment:
Replacing `Slice(start, stop, s_step).reduce(d).raw` caused overall runtime in the 1D use case to drop from 89ms to 42ms.